addons/mail/models/mail_scheduled_message.py

```diff
@@ -0,0 +1,255 @@
+            else:
+                raise UserError(_('You are not allowed to send this scheduled message'))
+        self._send()
+
+    def _send(self, raise_exception=True):
+        """ Send the scheduled messages
+            if raise_exception is set to False, the method will skip the posting of a message
+            instead of raising an error. This is useful when scheduled messages are sent from
+            the _send_messages_cron
+        """
+        messages = self.env['mail.message']
+        notification_parameters_whitelist = self._notification_parameters_whitelist()
+        for scheduled_message in self:
+            if raise_exception:
+                scheduled_message._check()
+            else:
+                try:
+                    scheduled_message._check()
+                except AccessError:
+                    _logger.info("Posting of scheduled message %s failed: user %s cannot post on the record", scheduled_message.id, scheduled_message.create_uid)
+                    continue
+            messages += self.env[scheduled_message.model].browse(scheduled_message.res_id).message_post(
+                attachment_ids=list(scheduled_message.attachment_ids.ids),
+                author_id=scheduled_message.author_id.id,
+                body=scheduled_message.body,
+                partner_ids=list(scheduled_message.partner_ids.ids),
+                subtype_xmlid='mail.mt_note' if scheduled_message.is_note else 'mail.mt_comment',
+                **{k: v for k, v in json.loads(scheduled_message.notification_parameters or '{}').items() if k in notification_parameters_whitelist},
+            )
+        self.unlink()  # should we handle the ones that failed and keep them with an error status?
+        return messages
+
+    # ------------------------------------------------------
+    # Business Methods
+    # ------------------------------------------------------
+
+    @api.model
+    def _check(self, values=None):
+        """ Restrict the access to a scheduled message.
+            Access is based on the record on which the scheduled message will be posted to
+        """
+        if self.env.is_superuser():
+            return True
+
+        model_ids = defaultdict(set)
+        if self:
+            for scheduled_message in self:
+                model_ids[scheduled_message.model].add(scheduled_message.res_id)
+        if values and (model := values.get('model')) and (res_id := values.get('res_id')):
+            model_ids[model].add(res_id)
+
+        for model, res_ids in model_ids.items():
+            records = self.env[model].browse(res_ids)
+            operation = records._mail_post_access
+            records.check_access(operation)
+
+    @api.model
+    def _notification_parameters_whitelist(self):
+        """ Parameters that can be used when posting the scheduled messages.
+        """
+        return {'email_from', 'force_email_lang', 'mail_auto_delete', 'mail_server_id', 'message_type'}
```

Comment:
Why so much restrictions ? e.g. email_layout_xmlid, reply_to, ... ? If we post using current user rights, shouldn't we consider that message_post is already sufficiently protected maybe ?